openage/opus/opusenc.pyx

```diff
@@ -0,0 +1,415 @@
+    # Search for 'data' chunk.
+    while wav[pos:pos+4] != b'data':
+        chk_len = int.from_bytes(wav[pos+4:pos+8], le)
+        pos += 8 + chk_len + (chk_len & 1) # Pad to even byte offset.
+        if pos >= len(wav):
+            return "Could not find 'data'-chunk in Wave-file."
+
+    chk_len = int.from_bytes(wav[pos+4:pos+8], le)
+    pos += 8
+    debug("read_wav: pos = {}",pos)
+    if len(wav) - pos < chk_len:
+        err = "Stated data-length {} in WAVE-header too high. " +\
+                "It should be at most {}. Is the file maybe truncated?"
+        return err.format(chk_len, len(wav) - pos)
+
+    inopt['header_len'] = pos
+    inopt['trailer_len'] = len(wav) - pos - chk_len
+    return inopt
+
+
+cdef write_ogg_page(offset, outdata, ogg_page *og):
+    '''
+    Write og at offset into outdata.
+    Returns new offset.
+    '''
+    data_len = og.header_len + og.body_len
+    outdata[offset:offset+og.header_len] =\
+        PyBytes_FromStringAndSize(<char*> og.header, og.header_len)
+    outdata[offset+og.header_len:offset+data_len] =\
+        PyBytes_FromStringAndSize(<char*> og.body, og.body_len)
+    return offset + data_len
+
+
+cdef void free_buffers(ogg_stream_state *os, OpusEncoder *oe,
+                       unsigned char *pymem1, unsigned char *pymem2):
+    ogg_stream_clear(os)
+    if oe != NULL:
+        opus_encoder_destroy(oe)
+    if pymem1 != NULL:
+        PyMem_Free(pymem1)
+    if pymem2 != NULL:
+        PyMem_Free(pymem2)
+    return
+
+
+cdef upsample(indata, opts, target_rate):
+    '''
+    Upsamples the PCM-data in 'indata' to 'target_rate' using linear
+    interpolation. 'opts' is a dict containing info on 'indata'.
+    If successful, return a byte object containing the upsampled data,
+    otherwise an error message.
+    Assumptions: * 1 ≤ opts['bit_depth'] ≤ 16
+                 * opts['input_rate'] < target_rate
+                 * opts['channels'] ∈ {1, 2}
+    TODO: * a more eloquent version?
+          '-> directly using bytearray instead of pointers?
+          * a clearer version?
+          * a faster version?
+          * a better interpolation?
+    '''
+    cdef:
+        int wframe_len = opts['wframe_len']
+        int channel = opts['channels']
+        int bit_depth = opts['bits_per_sample']
+        unsigned short mask = ~((1 << (16 - bit_depth)) - 1)
+
+        size_t iin = 0, iout = 0
+        size_t ismpls, osmpls
+        short a,b
+        int num, den
+
+        size_t outsize
+        const char *inp = indata
+        char *out
+
+    ismpls = len(indata) // wframe_len
+    if ismpls < 2:
+        # Nothing to do.
+        return indata
+
+    divisor = gcd(target_rate, opts['input_rate'])
+    num = target_rate / divisor
+    den = opts['input_rate'] / divisor
+
+    osmpls = ((ismpls - 1) * num /den).__trunc__() + 1 # asdf: explain
+    outsize = osmpls * wframe_len
+
+    out = <char *> PyMem_Malloc(outsize)
+    if out == NULL:
+        return "Out of memory."
```

Comment:
Otherwise that function would just return a error string, is that handled peoperly then? Python exceptions sound like the right way to abort after the error.